homeassistant/components/openweathermap/__init__.py

```diff
@@ -1 +1,103 @@
 """The openweathermap component."""
+import asyncio
+import logging
+
+from pyowm import OWM
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_API_KEY,
+    CONF_LATITUDE,
+    CONF_LONGITUDE,
+    CONF_MODE,
+    CONF_NAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import (
+    COMPONENTS,
+    CONF_LANGUAGE,
+    DOMAIN,
+    ENTRY_FORECAST_COORDINATOR,
+    ENTRY_NAME,
+    ENTRY_WEATHER_COORDINATOR,
+)
+from .forecast_update_coordinator import ForecastUpdateCoordinator
+from .weather_update_coordinator import WeatherUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up the OpenWeatherMap component."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Set up OpenWeatherMap as config entry."""
+    name = config_entry.data[CONF_NAME]
+    api_key = config_entry.data[CONF_API_KEY]
+    latitude = config_entry.data[CONF_LATITUDE]
+    longitude = config_entry.data[CONF_LONGITUDE]
+    forecast_mode = _get_config_value(config_entry, CONF_MODE)
+    language = _get_config_value(config_entry, CONF_LANGUAGE)
+
+    owm = OWM(API_key=api_key, language=language)
+    weather_coordinator = WeatherUpdateCoordinator(owm, latitude, longitude, hass)
+    forecast_coordinator = ForecastUpdateCoordinator(
+        owm, latitude, longitude, forecast_mode, hass
+    )
+
+    await weather_coordinator.async_refresh()
+    await forecast_coordinator.async_refresh()
+
+    if (
+        not weather_coordinator.last_update_success
+        and not forecast_coordinator.last_update_success
+    ):
+        raise ConfigEntryNotReady
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][config_entry.entry_id] = {
+        ENTRY_NAME: name,
+        ENTRY_WEATHER_COORDINATOR: weather_coordinator,
+        ENTRY_FORECAST_COORDINATOR: forecast_coordinator,
+    }
+
+    for component in COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    if not config_entry.update_listeners:
+        config_entry.add_update_listener(async_update_options)
+
```

Comment:
In the unload entry there must be an unsubscribe from the update listener

Reply:
a little help, how I can do that?

Reply:
Sure:

In `async_load_entry`
```
listener  = config_entry.add_update_listener(async_update_options)
hass.data[DOMAIN][entry.entry_id]["update_listener"] = listener
```

In 'async_unload_entry'
```
listener = hass.data[DOMAIN][entry.entry_id]["update_listener"]
listener()
```

I think you can remove the check `if not config_entry.update_listeners:` after doing this.

Reply:
@RobBie1221 could you please check out, is new version of this section is right

Reply:
Looks good! I'll have a look later at the rest as well